i18n: $localize message-id support

petebacondarwin opened the pull request:
~~Blocked on https://github.com/angular/angular/pull/32488~~

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

petebacondarwin commented on packages/localize/utils/BUILD.bazel:
```diff
@@ -12,6 +12,7 @@ ts_library(
         ],
     ),
     deps = [
+        "//packages/compiler",
```
A question, which will become more pertinent as the compile-time stuff comes in:

**What do we do about code that is needed in both `@angular/localize` and `@angular/compiler`?**

In this PR we need to share the `computeMsgId()` function between both packages. Later we will need to share XML parsing utilities.

1) I don't feel comfortable with either package depending upon the other.
2) Equally it is not a good idea to duplicate the code in both package sources.
3) It may be possible to do something clever with Bazel to have a single source Bazel package, which somehow gets bundled into both npm packages, but I have not worked out how to do this if it is possible.
4) Having yet another package (e.g. `@angular/compiler-utils`) is tempting but would add to the proliferation of dependencies in the Angular core system.

Currently I have gone with 1) making `@angular/localize` dependent upon `@angular/compiler`.

@mhevery and @IgorMinar  - any thoughts/ideas?

mhevery replied:
i like option #3. Could we do something with sim-links?

petebacondarwin replied:
I don't think symlinks work well with Bazel ...

AndrewKushnir reviewed (approved):
LGTM 👍 Great work @petebacondarwin!

cexbrayat commented on packages/localize/src/utils/constants.ts:
````diff
@@ -7,13 +7,37 @@
+ *
+ * ```ts
+ * $localize `:A welcome message on the home page@@myApp-homepage-welcome: Welcome!`;
+ */
````
nit: missing closing backticks

cexbrayat commented on packages/localize/src/utils/messages.ts:
```diff
@@ -53,29 +77,95 @@ export interface ParsedMessage {
+/**
+ * Parse the given message part (`cooked` + `raw`) to extract the message metadata from the text.
+ *
+ * If the message part has a metadata block contain up to three pieces of information (meaning,
```
nit: I think a word is missing here?

petebacondarwin committed:
feat(ivy): i18n - add syntax support for `$localize` metadata block

This commit documents and extends the basic `$localize`
implementation to support adding a metadata block to the
start of a tagged message.

All the "pass-though" version does is to strip this block out,
similar to what it does to placeholder name blocks.

petebacondarwin committed:
refactor(ivy): i18n - run-time translation API to use message-id

Previously the translation key used for translations was the `SourceMessage`
but it turns out that this is insufficient because "meaning" and "custom-id"
metadata affect the translation key.

Now run-time translation is keyed off the `MessageId`.

petebacondarwin committed:
test(ivy): i18n - re-enable skipped tests

The `packages/localize/test/utils` folder was not being
included in the unit tests because the glob for the spec
files was only looking in the top level folder.

petebacondarwin committed:
fix(ivy): i18n - update the compiler to output `MessageId`s

Now that the `$localize` translations are `MessageId` based the
compiler must render `MessageId`s in its generated `$localize` code.
This is because the `MessageId` used by the compiler is computed
from information that does not get passed through to the `$localize`
tagged string.

For example, the generated code for the following template

```html
<div id="static" i18n-title="m|d" title="introduction"></div>
```

will contain these localization statements

```ts
if (ngI18nClosureMode) {
  /**
    * @desc d
    * @meaning m
    */
  const MSG_EXTERNAL_8809028065680254561$$APP_SPEC_TS_1 = goog.getMsg("introduction");
  I18N_1 = MSG_EXTERNAL_8809028065680254561$$APP_SPEC_TS_1;
}
else {
  I18N_1 = $localize \`:m|d@@8809028065680254561:introduction\`;
}
```

Since `$localize` is not able to accurately regenerate the source-message
(and so the `MessageId`) from the generated code, it must rely upon the
`MessageId` being provided explicitly in the generated code.

The compiler now prepends all localized messages with a "metadata block"
containing the id (and the meaning and description if defined).

Note that this metadata block will also allow translation file extraction
from the compiled code - rather than relying on the legacy ViewEngine
extraction code. (This will be implemented post-v9).

Although these metadata blocks add to the initial code size, compile-time
inlining will completely remove these strings and so will not impact on
production bundle size.

petebacondarwin committed:
fix(ivy): i18n - use `MessageId` for matching translations

As discussed in https://hackmd.io/33M5Wb-JT7-0fneA0JuHPA `SourceMessage`
strings are not sufficient for matching translations.

This commit updates `@angular/localize` to use `MessageId`s for translation
matching instead.

Also the run-time translation will now log a warning to the console if a
translation is missing.

BREAKING CHANGE:

Translations (loaded via the `loadTranslations()` function) must now use
`MessageId` for the translation key rather than the previous `SourceMessage`
string.

AndrewKushnir commented:
FYI, VE and Ivy presubmits are successful.

IgorMinar commented on packages/localize/src/utils/messages.ts:
````diff
@@ -31,7 +51,7 @@ export type SourceMessage = string;
  * ```
  * {
- *   translationKey: 'Hello {$title}!',
+ *   messageId: 'Hello {$title}!',
````
is `'Hello {$title}!'` right?

petebacondarwin replied:
Good spot!

petebacondarwin commented:
@AndrewKushnir - I had to rebase because of a conflict with master. Do you need to run a new presubmit?

petebacondarwin commented:
@IgorMinar PTAL

AndrewKushnir commented:
@petebacondarwin if there are no major code changes, we can reuse previous TAP run. Thank you.

IgorMinar commented on packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts:
```diff
@@ -176,7 +176,7 @@ describe('i18n support in the template compiler', () => {
         <div i18n="meaningA|descA@@idA">Content A</div>
         <div i18n-title="meaningB|descB@@idB" title="Title B">Content B</div>
-        <div i18n-title="meaningC" title="Title C">Content C</div>
+        <div i18n-title="meaningC|" title="Title C">Content C</div>
```
ah.. I see, it is.

IgorMinar commented on packages/localize/src/utils/messages.ts:
```diff
@@ -53,29 +77,108 @@ export interface ParsedMessage {
+ * exists.
+ */
+export function splitBlock(cooked: string, raw: string): {text: string, block?: string} {
+  // There is a problem with synthesizing template literals using the TypeScript API.
```
it took me a while to understand what you meant here. maybe you could clarify the text. If I understand correctly then what you mean is something long the lines of: *creating AST nodes representing template literals using TypeScript apis is problematic because it doesn't allow for the raw value of `messageParts` to be programmatically set, leaving it empty.*

petebacondarwin replied:
Yeah - it is not so clear. Let me fix that up. Thanks!

IgorMinar reviewed (approved):
left a bunch of comments, but overall lgtm

I am however growing more and more concerned about the amount of code we generate (which will affect dev mode).

petebacondarwin commented:
> I am however growing more and more concerned about the amount of code we generate (which will affect dev mode).

@IgorMinar - I too am concerned about this. The problem we have is that it is not possible to recreate the computed message-ids from the information that is available at the time we call `$localize`.

This could be avoided if we were to make breaking changes to the digest algorithms used to compute these message-ids.  I believe that the spec (at least for XLIFF 1 and 2) does not specify the digest that should be used. The versions we use are specific to Angular - I believe somewhat based off some Google internal spec?? - so it is within our power to choose an algorithm that only cares about the message parts and placeholder names as available to `$localize`.

If we were to change our algorithms this would mean that all current message-id based translation files would be invalid and no longer match. It is feasible that we could provide a migration tool to move them to the new common algorithm.


By the way, you'll see (eventually once it is ready) that we have to do a whole load of processing in the compile-time inlining tool to convert these "external" message-ids to an "internal" format anyway for it to be able to work. I expect that fully functioning run-time translation that loads up translation files would need to do that same as well.

petebacondarwin commented on packages/localize/init/index.ts:
````diff
@@ -23,6 +23,22 @@ declare global {
+   * ```
+   *
+   * This format is the same as that used for `i18n` markers in Angular templates. See the
+   * [Angular 18n guide](guide/i18n#template-translations).
````
Let me double check...

petebacondarwin replied:
Take a look at this: https://angular.io/guide/i18n#help-the-translator-with-a-description-and-meaning

The code in this PR is correct. The format is `meaning|description@@id` and this is all metadata for the message itself which is specified either as the body of the element:

```html
<h1 i18n="meaning|description@@id">source message</h1>
```

or in a separate attribute:

```html
<h1 i18n-title="meaning|description@@id" title="source message">...</h1>
```

petebacondarwin commented on packages/localize/src/translate.ts:
```diff
@@ -6,25 +6,37 @@
  *
- * The given `translations` are processed and added to a lookup based on their translation key.
- * A new translation will overwrite a previous translation if it has the same key.
+ * The given `translations` are processed and added to a lookup based on their message-id.
```
In my comments I use the terms `message-id`, `source-message` and `target-message` to indicate i18n concepts, whereas I use `messageId`, `sourceMessage` and `targetMessage` when I want to refer to the specific TypeScript types in the code.

I can see that this distinction doesn't actually add much and opens up the opportunity for confusion, as you point out. So I will change it now.

petebacondarwin commented:
Merge-assistance: Presubmit should be fine according to @AndrewKushnir (https://github.com/angular/angular/pull/32594#issuecomment-531602949)

AndrewKushnir commented:
@petebacondarwin thanks for additional fixes. We've just landed big styling PR, so it'd be great to rebase to re-run CI with the most recent changes, could you please help with this?

I also started new TAP Presubmits:

- [VE Presubmit](https://test.corp.google.com/ui#id=OCL:269451494:BASE:269451520:1568677686759:5f5b2333)
- [Ivy Presubmit](https://test.corp.google.com/ui#id=OCL:269451494:BASE:269451642:1568677717149:b8a5b56)

If everything is "green", I'll proceed with merging this PR. Thank you.

AndrewKushnir commented:
FYI, VE and Ivy Presubmits look normal, CI is green after rebase, PR is ready to be merged. Thank you.
